File: design-patterns/ts/strategy/file-storage/storage-strategy.ts
// storage-strategy.ts

import StorageSystem from "./storage-system";

class StorageStrategy {
    private storage: StorageSystem;

    constructor(storage: StorageSystem) {
        this.storage = storage;
    }

    uploadFile(tempPath: string): number {
        const fileId = this.storage.storeFile(tempPath);
        this.storage.printFileInfo(fileId);

        return fileId;
    }

    getFileUrl(fileId: number): string {
        return this.storage.retrieveFile(fileId);
    }
}

export default StorageStrategy;

File: design-patterns/go/strategy/file_storage/storage_strategy.go
// storage_strategy.go

package main

type StorageStrategy struct {
	storage Storage
}

func NewStorageStrategy(storage Storage) (storageStrategy *StorageStrategy) {
	storageStrategy = &StorageStrategy{}
	storageStrategy.storage = storage
	return
}

func (storageStrategy *StorageStrategy) GetFileUrl(fileId int) string {
	return storageStrategy.storage.RetrieveFile(fileId)
}

func (storageStrategy *StorageStrategy) UploadFile(tempPath string) int {
	fileId := storageStrategy.storage.StoreFile(tempPath)
	storageStrategy.storage.PrintFileInfo(fileId)
	return fileId
}
